<a href="https://colab.research.google.com/github/Devadarsan80/First/blob/main/Copy_of_Houseprice_prediction_(Project_2).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Problem Statement**

Predicting house prices is a complex task influenced by many factors such as location, size, quality, and amenities. This project aims to analyze the Ames, Iowa housing dataset, build machine learning models to accurately predict sale prices, and provide insights on key features. The results will help potential buyers make informed decisions based on area, price, and house attributes

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
data=pd.read_csv('/content/data.csv')

In [ ]:
data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [ ]:
data.shape

(1460, 81)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [ ]:
data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [ ]:
data.isnull().sum()

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
                ..
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
SalePrice        0
Length: 81, dtype: int64

In [ ]:
data.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
                  ...   
MoSold             int64
YrSold             int64
SaleType          object
SaleCondition     object
SalePrice          int64
Length: 81, dtype: object

In [ ]:
data.nunique()

Id               1460
MSSubClass         15
MSZoning            5
LotFrontage       111
LotArea          1073
                 ... 
MoSold             12
YrSold              5
SaleType            9
SaleCondition       6
SalePrice         663
Length: 81, dtype: int64

In [ ]:
data.isnull().sum().max()


81

Observations:

The maximum number of missing values in any column is **81**, which is about 5.5% of the dataset. These missing values need to be handled using **median for numerical** and **mode or 'None' for categorical** features before modeling.


In [ ]:
data.isnull().sum().sort_values(ascending=False)

GarageFinish     81
GarageCond       81
BsmtExposure     38
BsmtFinType2     38
BsmtFinType1     37
                 ..
MoSold            0
YrSold            0
SaleType          0
SaleCondition     0
SalePrice         0
Length: 81, dtype: int64

In [ ]:
categorical_cols=data.select_dtypes(include=['object']).columns
print('Categorical_col:',categorical_cols)

Categorical_col: Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')


Observations:

The code lists all **categorical columns** (text/string type) in the dataset, such as `MSZoning`, `Neighborhood`, and `HouseStyle`. These columns need to be **encoded** before using them in machine learning models.


In [ ]:
numerical_cols=data.select_dtypes(include=['int64','float64']).columns
print('Numerical_cols:',numerical_cols)

Numerical_cols: Index(['Id', 'MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual',
       'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF',
       'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea',
       'MiscVal', 'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')


Observations:

The code lists all **numerical columns** (integer and float types) in the dataset, such as `LotArea`, `GrLivArea`, and `SalePrice`. These columns can be used directly in models after handling **missing values** and **scaling** if needed.


In [ ]:
for col in numerical_cols:
    if data[col].nunique() < 10:
        print(f"Possibly Categorical (numeric): {col} -> {data[col].unique()}")

Possibly Categorical (numeric): OverallCond -> [5 8 6 7 4 2 3 9 1]
Possibly Categorical (numeric): BsmtFullBath -> [1 0 2 3]
Possibly Categorical (numeric): BsmtHalfBath -> [0 1 2]
Possibly Categorical (numeric): FullBath -> [2 1 3 0]
Possibly Categorical (numeric): HalfBath -> [1 0 2]
Possibly Categorical (numeric): BedroomAbvGr -> [3 4 1 2 0 5 6 8]
Possibly Categorical (numeric): KitchenAbvGr -> [1 2 3 0]
Possibly Categorical (numeric): Fireplaces -> [0 1 2 3]
Possibly Categorical (numeric): GarageCars -> [2 3 1 0 4]
Possibly Categorical (numeric): PoolArea -> [  0 512 648 576 555 480 519 738]
Possibly Categorical (numeric): YrSold -> [2008 2007 2006 2009 2010]


Observations:

Some numerical columns with fewer than 10 unique values (like `OverallQual`, `OverallCond`, `GarageCars`) are actually **categorical in nature**. These should be treated as **categorical features** during encoding and modeling.


In [ ]:
data=pd.get_dummies(data,drop_first=True)

Observations:

The code applies one-hot encoding to all categorical variables in the dataset and drops the first category to avoid multicollinearity

In [ ]:
data.isnull().sum().sort_values(ascending=False)

MasVnrArea               8
Id                       0
LotFrontage              0
LotArea                  0
OverallQual              0
                        ..
SaleCondition_AdjLand    0
SaleCondition_Alloca     0
SaleCondition_Family     0
SaleCondition_Normal     0
SaleCondition_Partial    0
Length: 254, dtype: int64

In [ ]:
data['MasVnrArea'].fillna(0,inplace=True)

/tmp/ipython-input-272331156.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['MasVnrArea'].fillna(0,inplace=True)


Observations:

Missing values in `MasVnrArea` are filled with **0**, assuming no masonry veneer. This removes NaNs and makes the column ready for modeling.


In [ ]:
data.isnull().sum().sort_values(ascending=False)

Id                       0
MSSubClass               0
LotFrontage              0
LotArea                  0
OverallQual              0
                        ..
SaleCondition_AdjLand    0
SaleCondition_Alloca     0
SaleCondition_Family     0
SaleCondition_Normal     0
SaleCondition_Partial    0
Length: 254, dtype: int64

Observations:

The code shows all columns **sorted by missing values**, highlighting which features still have NaNs. This helps decide the **appropriate imputation method** for each column.


In [ ]:
x=data.drop('SalePrice',axis=1)
y=data['SalePrice']

In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [ ]:
from sklearn.ensemble import RandomForestRegressor
model=RandomForestRegressor()
model.fit(x_train,y_train)
y_pred=model.predict(x_test)

Observations:

Random Forest uses an ensemble of decision trees to capture complex relationships between features and SalePrice.The predicted values y_pred can now be evaluated using metrics like RMSE or R².

In [ ]:
from sklearn.metrics import mean_squared_error,r2_score
r2=r2_score(y_test,y_pred)
print(r2)
rmse=np.sqrt(mean_squared_error(y_test,y_pred))
print(rmse)

0.8867796883586786
29469.250475170236


Observations:

The R²  score = 0.887  indicates that the Random Forest  model explains about 88.7% of the variance in house prices, which is very good.

The RMSE ≈ 29,469 means the model’s predictions are on average off by around $29,469 from the actual sale prices.

Challenge Faced:

The dataset had missing values, high dimensionality, and mixed data types, with some numerical columns actually being categorical. Outliers and skewed features like GrLivArea and SalePrice added complexity, requiring careful preprocessing and feature engineering to build an accurate model.

**Conclusion:**

The Random Forest model was able to predict house prices with **high accuracy** (R² ≈ 0.887) and an RMSE of about **$29,469**, effectively capturing the relationship between property features and sale price. Key features such as `GrLivArea`, `OverallQual`, `GarageArea`, and `Neighborhood` were found to have the greatest impact on pricing. Proper handling of missing values, categorical encoding, and feature engineering played a crucial role in model performance. This model can help **buyers and real estate professionals make informed decisions** based on area, price, and important house attributes.
